Jacob Valdenegro A01640992

Roberto Rillo A01642022

Verifiar los resultados utilizando Python.

In [19]:
import numpy as np

X = np.array([[1, 2],
              [1, 3],
              [1, 5],
              [1, 7],
              [1, 9]])

# Calcular la matriz HAT
H = X @ np.linalg.inv(X.T @ X) @ X.T

# Valores reales de y
y = np.array([5, 8, 7, 10, 12])

# Valores predichos: ŷ = H * y
y_pred = H @ y

# Coeficientes de regresión: β = (X^T X)^(-1) X^T y
beta = np.linalg.inv(X.T @ X) @ X.T @ y

# Imprimir resultados
print("Matriz HAT (H):\n", H)
print("\nValores predichos (ŷ):\n", y_pred)
print("\nCoeficientes de la regresión (β):\n", beta)


Matriz HAT (H):
 [[ 0.51219512  0.41463415  0.2195122   0.02439024 -0.17073171]
 [ 0.41463415  0.34756098  0.21341463  0.07926829 -0.05487805]
 [ 0.2195122   0.21341463  0.20121951  0.18902439  0.17682927]
 [ 0.02439024  0.07926829  0.18902439  0.29878049  0.40853659]
 [-0.17073171 -0.05487805  0.17682927  0.40853659  0.6402439 ]]

Valores predichos (ŷ):
 [ 5.6097561   6.48170732  8.22560976  9.9695122  11.71341463]

Coeficientes de la regresión (β):
 [3.86585366 0.87195122]


Utilizando los Datos “Cirugía de Hígado” obtener la matriz HAT, los valores predichos, y ̂, y los coeficientes de la ecuación de regresión β utilizando el método de matrices.

In [ ]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
import numpy as np

# Cargar el archivo Excel
archivo = 'Covarianza, Correlación y HAT.xlsx'
datos = pd.read_excel(archivo)

# Verifica que los datos se cargaron correctamente
print(datos.head())  # Muestra las primeras filas


   Factor Coagulación  Índice pronóstico  Función de enzima  \
0                 6.7                 62                 81   
1                 5.1                 59                 66   
2                 7.4                 57                 83   
3                 6.5                 73                 41   
4                 7.8                 65                115   

   Función de hígado  Edad  Género  Alcohol (moderado)  Alcohol (severo)  \
0               2.59    50       0                   1                 0   
1               1.70    39       0                   0                 0   
2               2.16    55       0                   0                 0   
3               2.01    48       0                   0                 0   
4               4.30    45       0                   0                 1   

   Sobrevivencia (días)  
0                   695  
1                   403  
2                   710  
3                   349  
4                  2343  


In [ ]:
# Variables predictoras: todas menos 'Sobrevivencia (días)'
X = datos.drop(columns=['Sobrevivencia (días)'])

# Variable objetivo
y = datos['Sobrevivencia (días)']

# Escalar las variables predictoras
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

# Convertir a matriz NumPy
X_scaled = np.array(X_scaled)
y = np.array(y)


In [ ]:
X_scaled = np.hstack([np.ones((X_scaled.shape[0], 1)), X_scaled])

# Calcular la matriz HAT
XTX_inv = np.linalg.inv(X_scaled.T @ X_scaled)
H = X_scaled @ XTX_inv @ X_scaled.T

# Calcular los coeficientes β
beta = XTX_inv @ X_scaled.T @ y

# Calcular los valores predichos
y_pred = X_scaled @ beta

# Valores de apalancamiento (diagonal de la matriz HAT)
apalancamiento = np.diag(H)


In [ ]:
# Imprimir resultados
print("\nCoeficientes de la regresión (β):")
print(beta)

print("\nValores predichos (ŷ):")
print(y_pred)

print("\nValores de apalancamiento:")
print(apalancamiento)

print("\nMatriz HAT:")
print(H)


Coeficientes de la regresión (β):
[-575.86374747  453.26085291  738.23244281  854.14304366  429.43164879
   25.63892773   13.09258075  -41.26764482  195.70703222]

Valores predichos (ŷ):
[ 706.25623722  430.82292124  732.22983195  425.03957611 1454.587552
  317.98512422  561.43434506  620.57029786  814.99238167  761.68391557
  989.9519364   228.02181205 1394.13461057 1050.29172024  937.95668893
  732.48434715  418.09976441   58.32293366  857.99665125  939.03472439
  555.53153245  366.02513011  485.52130235  817.77817446  861.31877768
  669.35808718  592.75876392 1619.88673751  399.48541635  596.03964243
  257.21831085  163.93595245  594.67195134 1143.59947369  512.12738287
  702.86128551  629.09497899  563.49249113  556.62971999  607.66774824
  423.46168884  569.50576666 1363.25897377  625.40911054  688.74643782
  590.69719298  948.046226   1333.78069235  712.00261732 1065.38071841
  462.70846831  649.64082214  690.30288166  958.05440017  330.98090005
  811.81615789  635.24982009  367

La suma de los valores de apalancamiento (o leverages) es igual al número de coeficientes de la regresión en un modelo de regresión lineal, estos valores tambien reflejan qué tan influyentes son en la estimación de los coeficientes de regresión.

La ecuación de regresión seria:

ŷ = -575.86 * X1 + 453.26 * X2 + 738.23 * X3 + 854.14 * X4 + 429.43 * X5
     + 25.64 * X6 + 13.09 * X7 - 41.27 * X8 + 195.71